### PR TITLE
add psis diagnostic

### PR DESCRIPTION
**Pull request by martinjankowiak**
follow-up to #1800 

adds a VI diagnostic along the lines of what's described in [Yes, but Did It Work?](https://arxiv.org/abs/1802.02538)

enabled by #1804

### CODE REVIEW

File: pyro/infer/importance.py [hunk 1]
```diff
@@ -140,3 +145,79 @@ def _fn(*args, **kwargs):
     if normalized:
         log_weights = log_weights - torch.logsumexp(log_weights)
     return log_weights, model_trace, guide_trace
+
+
+def psis_diagnostic(model, guide, *args, **kwargs):
```
Comment (martinjankowiak) [inline pyro/infer/importance.py]: this is in here to avoid circular imports

File: pyro/ops/stats.py [hunk 2]
```diff
@@ -386,8 +386,9 @@ def fit_generalized_pareto(X):
     L = np.log(-bs / ks) - (ks + 1)
     L *= N
 
-    temp = np.exp(L - L[:, None])
-    w = 1.0 / np.sum(temp, axis=1)
+    L_max = np.max(L)
```
Comment (martinjankowiak) [inline pyro/ops/stats.py]: for additional numerical stability

**Comment by martinjankowiak** [comment]
@stefanwebb this should now be ready for prime time, although you'll probably want to use large numbers of particles (e.g. 10^6) to get accurate estimates

File: tests/infer/test_util.py [hunk 3]
```diff
@@ -44,3 +46,22 @@ def test_multi_frame_tensor():
     for name, expected_sum in expected.items():
         actual_sum = actual.sum_to(stacks[name])
         assert_equal(actual_sum, expected_sum, msg=name)
+
+
+@pytest.mark.parametrize('max_particles', [250 * 1000, 500 * 1000])
+@pytest.mark.parametrize('scale_krange', zip([0.50, 0.95], [(0.7, 0.9), (0.05, 0.2)]))
```
Comment (fritzo) [inline tests/infer/test_util.py]: Note you can parametrize multiple zipped names:
```py
@pytest.mark.parametrize('scale,krange', [(0.50, (0.79, 0.9)), (0.95, (0.05, 0.2))])
def test_psis_diagnostic(scale, krange, ...):
    ...
```

File: pyro/infer/importance.py [hunk 4]
```diff
@@ -140,3 +145,78 @@ def _fn(*args, **kwargs):
...
+
+    Finally note that this implementation is adapted from the implementation
+    in reference [2].
+
+    References
+    [1] 'Yes, but Did It Work?: Evaluating Variational Inference.'
+    Yuling Yao, Aki Vehtari, Daniel Simpson, Andrew Gelman
+    [2] 'Pareto Smoothed Importance Sampling.'
+    Aki Vehtari, Andrew Gelman, Jonah Gabry
+    https://github.com/avehtari/PSIS/blob/904146236767182270c9718a7b2a30831fe701fe/py/psis.py#L112
```
Comment (fritzo) [inline pyro/infer/importance.py]: Sorry, you can't adapt that code because it is GPL-3.0 licensed.

Comment (fritzo) [reply pyro/infer/importance.py]: Here's an Apache-2.0 licensed implementation of Pareto something, from arviz: https://github.com/arviz-devs/arviz/blob/master/arviz/stats/diagnostics.py#L331

**Pull request closed without merging**